File: backend/src/main/kotlin/com/trackranker/service/TrackService.kt
package com.trackranker.service

import com.trackranker.model.Track
import com.trackranker.repository.TrackRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class TrackService(
    private val trackRepository: TrackRepository
) {
    
    fun getAllTracks(): List<Track> = trackRepository.findAll()
    
    fun getTrackById(id: Long): Track? = trackRepository.findById(id).orElse(null)
    
    fun getTracksByState(state: String): List<Track> = trackRepository.findByState(state)
    
    fun searchTracks(query: String): List<Track> {
        return trackRepository.findByNameContainingIgnoreCaseOrCityContainingIgnoreCaseOrStateContainingIgnoreCase(
            query, query, query
        )
    }
    
    fun getTracksByDifficultyRange(minRating: Int, maxRating: Int): List<Track> {
        return trackRepository.findByDifficultyRange(minRating, maxRating)
    }
    
    fun createTrack(track: Track): Track = trackRepository.save(track)
    
    fun updateTrack(id: Long, track: Track): Track? {
        val existingTrack = trackRepository.findById(id).orElse(null) ?: return null
        val updatedTrack = existingTrack.copy(
            name = track.name,
            city = track.city,
            state = track.state,
            zipCode = track.zipCode,
            trackLength = track.trackLength,
            trackWidth = track.trackWidth,
            numberOfTurns = track.numberOfTurns,
            surfaceType = track.surfaceType,
            trackDirection = track.trackDirection,
            yearBuilt = track.yearBuilt,
            websiteUrl = track.websiteUrl,
            phoneNumber = track.phoneNumber,
            emailAddress = track.emailAddress,
            maxElevationChange = track.maxElevationChange,
            bankingAngles = track.bankingAngles,
            runOffAreas = track.runOffAreas,
            pitLaneLength = track.pitLaneLength,
            numberOfPitStalls = track.numberOfPitStalls,
            trackRecordCar = track.trackRecordCar,
            trackRecordMotorcycle = track.trackRecordMotorcycle,
            trackLayoutImageUrl = track.trackLayoutImageUrl,
            satelliteImageUrl = track.satelliteImageUrl,
            trackDayOrganizations = track.trackDayOrganizations,
            typicalPricing = track.typicalPricing,
            availableDates = track.availableDates,
            trackDaySchedule = track.trackDaySchedule,
            requiredSafetyEquipment = track.requiredSafetyEquipment,
            noiseRestrictions = track.noiseRestrictions,
            fuelAvailability = track.fuelAvailability,
            amenities = track.amenities,
            aiDifficultyRating = track.aiDifficultyRating,
            aiDifficultyExplanation = track.aiDifficultyExplanation,
            aiAssessmentDate = track.aiAssessmentDate,
            averageUserRating = track.averageUserRating,
            numberOfReviews = track.numberOfReviews,
            updatedAt = java.time.LocalDateTime.now()
        )
        return trackRepository.save(updatedTrack)
    }
    
    fun deleteTrack(id: Long): Boolean {
        return if (trackRepository.existsById(id)) {
            trackRepository.deleteById(id)
            true
        } else {
            false
        }
    }
} 

File: frontend/src/app/services/track.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { Track, TrackReview } from '../models/track.model';
import { TrackImageService } from './track-image.service';

@Injectable({
  providedIn: 'root'
})
export class TrackService {
  // Local development
  // private apiUrl = 'http://localhost:8080/api/tracks';
  
  // Production
  private apiUrl = 'https://track-ranker-backend-gtgdg6dxawhzgra9.centralus-01.azurewebsites.net/api/tracks';

  constructor(
    private http: HttpClient,
    private trackImageService: TrackImageService
  ) { }

  getTracks(): Observable<Track[]> {
    return this.http.get<any[]>(this.apiUrl).pipe(
      map(tracks => tracks.map(track => this.mapTrackFromBackend(track))),
      map(tracks => tracks.map(track => this.addSatelliteImage(track)))
    );
  }

  getTrack(id: number): Observable<Track> {
    return this.http.get<any>(`${this.apiUrl}/${id}`).pipe(
      map(track => this.mapTrackFromBackend(track)),
      map(track => this.addSatelliteImage(track))
    );
  }

  // Review methods
  getTrackReviews(trackId: number): Observable<TrackReview[]> {
    return this.http.get<TrackReview[]>(`${this.apiUrl}/${trackId}/reviews`);
  }

  submitReview(trackId: number, review: TrackReview): Observable<TrackReview> {
    return this.http.post<TrackReview>(`${this.apiUrl}/${trackId}/reviews`, review);
  }

  updateReview(reviewId: number, review: TrackReview): Observable<TrackReview> {
    return this.http.put<TrackReview>(`${this.apiUrl}/reviews/${reviewId}`, review);
  }

  deleteReview(reviewId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/reviews/${reviewId}`);
  }

  private mapTrackFromBackend(backendTrack: any): Track {
    return {
      id: backendTrack.id,
      name: backendTrack.name,
      city: backendTrack.city,
      state: backendTrack.state,
      zipCode: backendTrack.zipCode,
      trackLength: backendTrack.trackLength,
      trackWidth: backendTrack.trackWidth,
      numberOfTurns: backendTrack.numberOfTurns,
      surfaceType: backendTrack.surfaceType,
      trackDirection: backendTrack.trackDirection,
      yearBuilt: backendTrack.yearBuilt,
      websiteUrl: backendTrack.websiteUrl,
      phoneNumber: backendTrack.phoneNumber,
      emailAddress: backendTrack.emailAddress,
      maxElevationChange: backendTrack.maxElevationChange,
      bankingAngles: backendTrack.bankingAngles,
      runOffAreas: backendTrack.runOffAreas,
      pitLaneLength: backendTrack.pitLaneLength,
      numberOfPitStalls: backendTrack.numberOfPitStalls,
      trackRecordCar: backendTrack.trackRecordCar,
      trackRecordMotorcycle: backendTrack.trackRecordMotorcycle,
      trackLayoutImageUrl: backendTrack.trackLayoutImageUrl,
      satelliteImageUrl: backendTrack.satelliteImageUrl,
      aiDifficultyRating: Number(backendTrack.aiDifficultyRating),
      aiDifficultyExplanation: backendTrack.aiDifficultyExplanation,
      averageUserRating: Number(backendTrack.averageUserRating),
      numberOfReviews: Number(backendTrack.numberOfReviews),
      amenities: backendTrack.amenities,
      availableDates: backendTrack.availableDates,
      fuelAvailability: backendTrack.fuelAvailability,
      noiseRestrictions: backendTrack.noiseRestrictions,
      requiredSafetyEquipment: backendTrack.requiredSafetyEquipment,
      trackDayOrganizations: backendTrack.trackDayOrganizations,
      trackDaySchedule: backendTrack.trackDaySchedule,
      typicalPricing: backendTrack.typicalPricing,
      latitude: Number(backendTrack.latitude),
      longitude: Number(backendTrack.longitude)
    };
  }

  private addSatelliteImage(track: Track): Track {
    // Generate satellite image for tracks with coordinates
    if (track.latitude && track.longitude && track.latitude !== 0 && track.longitude !== 0) {
      track.satelliteImageUrl = this.trackImageService.getSatelliteImage(
        track.latitude, 
        track.longitude
      );
    } else {
      track.satelliteImageUrl = this.trackImageService.getSatelliteImage();
    }
    return track;
  }
} 